auditlog/iter.go

```diff
@@ -0,0 +1,67 @@
+package auditlog
+
+import (
+	"reflect"
+)
+
+// Query is the function used to get a page listing.
+type Query func(ListRequestParams) ([]interface{}, ListMeta, error)
+
+// EventItr represents and iterative event
+type Iter struct {
+	values  []interface{}
+	current interface{}
+	err     error
+	params  ListRequestParams
+	meta    ListMeta
+	query   Query
+}
+
+// Next returns the next event in the set
+func (it *Iter) Next() bool {
+	if len(it.values) == 0 && it.meta.HasMore {
+		if it.params.GetStartingAfter() != "" {
+			it.params.After = listItemID(it.current)
+		} else {
+			it.params.Before = listItemID(it.current)
+		}
+		it.getPage()
+	}
+
+	if len(it.values) == 0 {
+		return false
+	}
+
+	it.current = it.values[0]
+	it.values = it.values[1:]
+
+	return true
+}
+
+func (it Iter) Current() interface{} {
+	return it.current
+}
+
+func (it *Iter) getPage() {
+	it.values, it.meta, it.err = it.query(it.params)
+}
+
+func (it Iter) Err() error {
+	return it.err
+}
+
+// GetIter returns a new Iter for a given query and its options.
+func GetIter(params ListRequestParams, query Query) *Iter {
+	iter := &Iter{
+		params: params,
+		query:  query,
+	}
+
+	iter.getPage()
+
+	return iter
+}
+
+func listItemID(x interface{}) string {
```

Comment:
I think it'd make more sense to take advantage of go's interfaces so each each list item would have to implement the following function:

```go
type Iterable interface {
  GetID() string
}
```